File: services/controllers/link.go
package controllers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"slimlink/models"
	"slimlink/services/api"
	"strings"
)

func AddLinkHandler(w http.ResponseWriter, r *http.Request) {
	var linkRequestDTO models.LinkRequestDTO
	err := json.NewDecoder(r.Body).Decode(&linkRequestDTO)
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}
	link, err := api.AddLink(&linkRequestDTO)
	if err != nil {
		if err.Error() == "invalid URL" {
			w.WriteHeader(http.StatusBadRequest)
		} else {
			fmt.Println(err)
			w.WriteHeader(http.StatusInternalServerError)
		}
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(link)
}

func GetLinkHandler(w http.ResponseWriter, r *http.Request) {
	urlPathSplit := strings.Split(r.URL.Path, "/")[1:]
	link, err := api.GetLinkByID(urlPathSplit[2])
	if err != nil {
		if err.Error() == "link not found" {
			w.WriteHeader(http.StatusNotFound)
		} else {
			fmt.Println(err)
			w.WriteHeader(http.StatusInternalServerError)
		}
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(link)
}

func RedirectToLinkUrlHandler(w http.ResponseWriter, r *http.Request) {
	urlPathSplit := strings.Split(r.URL.Path, "/")[1:]
	link, err := api.GetLinkByID(urlPathSplit[2])
	if err != nil {
		if err.Error() == "link not found" {
			http.Redirect(w, r, "/404", http.StatusTemporaryRedirect)
		} else {
			fmt.Println(err)
			w.WriteHeader(http.StatusInternalServerError)
		}
		return
	}
	http.Redirect(w, r, link.Url, http.StatusTemporaryRedirect)
}

File: ui/src/services/server/controllers/link.ts
import type { NextApiRequest, NextApiResponse } from "next";
import { addLink, getLinkByID } from "services/server/api/link";
import { Link, LinkRequestDTO } from "types";

export async function addLinkHandler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
	if (req.method !== "POST") {
		res.status(405).end();
		return;
	}
	const linkRequestDTO: LinkRequestDTO = req.body;
	let link = null;
	try {
		link = await addLink(linkRequestDTO);
	} catch {
		res.status(400).end();
		return;
	}
	res.status(200).json(link);
}

async function runWithLink(req: NextApiRequest, res: NextApiResponse, func: (link: Link | null) => void): Promise<void> {
	if (req.method !== "GET") {
		res.status(405).end();
		return;
	}
	if (typeof req.query.id !== "string") {
		res.status(400).end();
		return;
	}
	let link = null;
	try {
		link = await getLinkByID(req.query.id);
	} catch {}
	func(link);
}

export async function getLinkHandler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
	await runWithLink(req, res, (link) => {
		if (!link) {
			res.status(404).end();
			return;
		}
		res.status(200).json(link);
	});
}

export async function redirectToLinkUrlHandler(req: NextApiRequest, res: NextApiResponse): Promise<void> {
	await runWithLink(req, res, (link) => {
		if (!link) {
			res.redirect("/404");
			return;
		}
		res.redirect(link.url);
	});
}
